Index: linux-2.6.36/arch/mips/Kconfig
===================================================================
--- linux-2.6.36.orig/arch/mips/Kconfig
+++ linux-2.6.36/arch/mips/Kconfig
@@ -45,7 +45,7 @@ config RALINK_RT2880
 	select SWAP_IO_SPACE
 	select SYS_HAS_CPU_MIPS32_R1
 	select SYS_SUPPORTS_LITTLE_ENDIAN
-	select BOOT_ELF32
+	select BOOT_RAW
 	select SYS_SUPPORTS_32BIT_KERNEL
 	select RALINK_RT2880_MP
 	select CSRC_R4K
@@ -62,7 +62,7 @@ config RALINK_RT3052
         select SYS_HAS_CPU_MIPS32_R2
 	select SYS_HAS_EARLY_PRINTK
 	select SYS_SUPPORTS_LITTLE_ENDIAN
-	select BOOT_ELF32
+	select BOOT_RAW
 	select SYS_SUPPORTS_32BIT_KERNEL
 	select DEBUG_INFO
 	select CSRC_R4K
@@ -79,7 +79,7 @@ config RALINK_RT3352
 	select SYS_HAS_CPU_MIPS32_R1
         select SYS_HAS_CPU_MIPS32_R2
 	select SYS_SUPPORTS_LITTLE_ENDIAN
-	select BOOT_ELF32
+	select BOOT_RAW
 	select SYS_SUPPORTS_32BIT_KERNEL
 	select RALINK_RT3352_MP
 	select CSRC_R4K
@@ -95,7 +95,7 @@ config RALINK_RT5350
 	select SYS_HAS_CPU_MIPS32_R1
 	select SYS_HAS_CPU_MIPS32_R2
 	select SYS_SUPPORTS_LITTLE_ENDIAN
-	select BOOT_ELF32
+	select BOOT_RAW
 	select SYS_SUPPORTS_32BIT_KERNEL
 	select RALINK_RT5350_MP
 	select CSRC_R4K
@@ -112,7 +112,7 @@ config RALINK_RT3883
 	select SYS_HAS_CPU_MIPS32_R1
         select SYS_HAS_CPU_MIPS32_R2
 	select SYS_SUPPORTS_LITTLE_ENDIAN
-	select BOOT_ELF32
+	select BOOT_RAW
 	select SYS_SUPPORTS_32BIT_KERNEL
 	select RALINK_RT3883_MP
 	select CSRC_R4K
@@ -129,7 +129,7 @@ config RALINK_MT7620
 	select SYS_HAS_CPU_MIPS32_R1
         select SYS_HAS_CPU_MIPS32_R2
 	select SYS_SUPPORTS_LITTLE_ENDIAN
-	select BOOT_ELF32
+	select BOOT_RAW
 	select SYS_SUPPORTS_32BIT_KERNEL
 	select CSRC_R4K
 	select CEVT_R4K
@@ -145,7 +145,7 @@ config RALINK_MT7621
 	select SYS_HAS_CPU_MIPS32_R1
         select SYS_HAS_CPU_MIPS32_R2
 	select SYS_SUPPORTS_LITTLE_ENDIAN
-	select BOOT_ELF32
+	select BOOT_RAW
 	select SYS_SUPPORTS_32BIT_KERNEL
 	select CPU_SUPPORTS_HIGHMEM
 	select SYS_SUPPORTS_HIGHMEM
@@ -168,7 +168,7 @@ config RALINK_MT7628
 	select SYS_HAS_CPU_MIPS32_R1
         select SYS_HAS_CPU_MIPS32_R2
 	select SYS_SUPPORTS_LITTLE_ENDIAN
-	select BOOT_ELF32
+	select BOOT_RAW
 	select SYS_SUPPORTS_32BIT_KERNEL
 	select CSRC_R4K
 	select CEVT_R4K
@@ -522,7 +522,7 @@ config ARC32
 config SNIPROM
 	bool
 
-config BOOT_ELF32
+config BOOT_RAW
 	bool
 
 config MIPS_L1_CACHE_SHIFT
